### PR TITLE
Memorialize borrower threshold price on kick 

### PR DESCRIPTION
**Pull request by prat-gpt**
## Description

* Memorialize borrower threshold price at the time of kick to calculate bpf.
## Purpose

* The kicker penalty was getting reduced with subsequent takes due to a decrease in threshold price with each take. To avoid this situation, threshold price is stored at the time of kick and the same used to calculate bpf
* Solves the issue discovered in Sherlock audit -> https://github.com/sherlock-audit/2023-09-ajna-judging/issues/69

## Impact

Small increase in gas cost of kicks and takes functions

## Tasks

- [x] Changes to protocol contracts are covered by unit tests executed by CI.
- [x] Protocol contract size limits have not been exceeded.
- [x] Gas consumption for impacted transactions have been compared with the target branch, and nontrivial changes cited in the _Impact_ section above.
- [x] Scope labels have been assigned as appropriate.
- [x] Invariant tests have been manually executed as appropriate for the nature of the change.

### CODE REVIEW

**Review by EdNoepel — COMMENTED** [review]
Protocol changes look good.  After other PRs have been merged, we can create a merge branch and change target of this to the merge branch.

File: src/libraries/external/KickerActions.sol
Comment (mattcushman) [inline src/libraries/external/KickerActions.sol]: I think this would be more logically put in `_recordAuction` perhaps?

**Review by mattcushman — COMMENTED** [review]
i think it's fine functionally, but for readability and localization I feel that memorializing the TP would be better done inside `_recordAuction` alongside other vars like the NP?

File: src/libraries/external/KickerActions.sol
Comment (prat-gpt) [reply src/libraries/external/KickerActions.sol]: I thought there will be a increase in gas cost if we do inside `_recordAuction`, but now when I checked the increase in gas and it is a very small amount. I think for readability in this case we can do it inside `_recordAuction`. I will update.

File: src/libraries/helpers/PoolHelper.sol
Comment (ith-harvey) [inline src/libraries/helpers/PoolHelper.sol]: Does it make sense to wrap `int256(thresholdPrice_)` here and on line 396? or can we just wrap it once at the begginging of the method logic?

File: src/libraries/external/KickerActions.sol
Comment (prat-gpt) [reply src/libraries/external/KickerActions.sol]: updated [7710e22](https://github.com/ajna-finance/contracts/pull/976/commits/7710e224c1bc154e44246388bd4bfb889db98e25)

**Review by ith-harvey — COMMENTED** [review]
Waiting for the stamping of the `thresholdPrice` to be moved into `_recordAuction` before I approve

**Comment by prat-gpt** [comment]
> Waiting for the stamping of the `thresholdPrice` to be moved into `_recordAuction` before I approve

done here already [7710e22](https://github.com/ajna-finance/contracts/pull/976/commits/7710e224c1bc154e44246388bd4bfb889db98e25) @ith-harvey

File: src/libraries/helpers/PoolHelper.sol
Comment (prat-gpt) [reply src/libraries/helpers/PoolHelper.sol]: we don't need int conversion in the if block. fixing this.

Comment (prat-gpt) [reply src/libraries/helpers/PoolHelper.sol]: done here [1aed6df](https://github.com/ajna-finance/contracts/pull/976/commits/1aed6dfeb52c2f875ed365e1dcd4b58bf4e0a2ee)

**Review by ith-harvey — APPROVED** [review]
LGTM once conflicts are handled

**Review by mattcushman — APPROVED** [review]

**Review by EdNoepel — APPROVED** [review]
Surprised there wasn't more test impact.

**Review by dmitriia — COMMENTED** [review]
Looks ok

**Pull request merged**
